Modules/_sqlite/connection.c

```diff
@@ -1074,11 +1073,37 @@ trace_callback(void *ctx, const char *statement_string)
 
     PyGILState_STATE gilstate = PyGILState_Ensure();
 
+    assert(ctx != NULL);
     PyObject *py_statement = NULL;
+#ifdef HAVE_TRACE_V2
+    assert(stmt != NULL);
+    const char *expanded_sql = sqlite3_expanded_sql((sqlite3_stmt *)stmt);
+    if (expanded_sql == NULL) {
+        sqlite3 *db = sqlite3_db_handle((sqlite3_stmt *)stmt);
+        if (sqlite3_errcode(db) == SQLITE_NOMEM) {
+            (void)PyErr_NoMemory();
+            goto exit;
+        }
+
+        pysqlite_state *state = ((callback_context *)ctx)->state;
+        assert(state != NULL);
+        if (state->enable_callback_tracebacks) {
+            PyErr_SetString(state->DataError,
+                            "Expanded SQL string exceeds the maximum string "
+                            "length");
+            PyErr_Print();
```

Comment:
Why this? If this is an unraisable exception you may want to use `PyErr_WriteUnraisable`.

Reply:
Ah, I was unaware of that C API function. Nice; thanks!

We've never raised exceptions in the trace handler. I had a discussion with Petr regarding this on https://github.com/python/cpython/pull/28133#issuecomment-914286495.

Also, `sqlite3` has this "print an exception or clear it" option before signalling UDF errors back to SQLite.

Reply:
FYI, I've created a WIP branch that modifies`print_or_clear_traceback()` to use unraisable exceptions. I prefer to land this PR using the old/current behaviour before possibly changing the callback code to use unraisable exceptions.

Reply:
I'm marking this conversation as resolved. Feel free to undo if you disagree :)

Reply:
I opened [bpo-458282](https://bugs.python.org/issue45828) and GH-29591 for this.